File: Cpp/CXR_classify/downloader.cpp
#include "downloader.h"

#include <chrono>
#include <iostream>

namespace fs = std::filesystem;

Downloader::Downloader(ConfigHandler * configHandler) : Runnable(configHandler)
{
    filenameAbsPath = configHandler->getTgzFilePath();
    datasetType = configHandler->getDatasetType();
}

void Downloader::run()
{
    // std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    logger->info("Checking if {} already exists", filenameAbsPath);
    if (fs::exists(filenameAbsPath) && !fs::is_directory(filenameAbsPath)) {
        logger->info("{} already exists", filenameAbsPath);
        logger->info("Checking if {} was downloaded properly", filenameAbsPath);
        if (fs::file_size(filenameAbsPath) == expected_size) {
            logger->info("{} was downloaded properly", filenameAbsPath);
        } else {
            logger->warn("{} was not downloaded properly", filenameAbsPath);
            logger->info("Removing {}", filenameAbsPath);
            fs::remove(filenameAbsPath);
            logger->info("Successfully removed {}", filenameAbsPath);
            download();
        }
    } else {
        logger->info("{} does not exist", filenameAbsPath);
        download();
    }
    // std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    // std::cout << "Time difference = " << (std::chrono::duration_cast<std::chrono::nanoseconds> (end - begin).count())/1000000000.0 << "[s]" << std::endl;
    emit attemptUpdateProBarValue(getTgzSize());
    emit attemptUpdateText("Image download complete");
    emit finished();
}

int Downloader::download()
{
    logger->info("Downloading dataset from {}", configHandler->getUrl());

    emit attemptUpdateText("Downloading images");
    emit attemptUpdateProBarBounds(0, getTgzMax());

    QNetworkAccessManager nam;
    QFile file(filenameAbsPath.c_str());
    if(!file.open(QIODevice::ReadWrite)) {
        // log "Can't open write file"
    }
    QNetworkRequest request(QUrl(configHandler->getUrl().c_str()));
    QNetworkReply* reply = nam.get(request);
    QEventLoop event;

    QObject::connect(reply, &QNetworkReply::readyRead, [&]{
        //this will be called every time a chunk of data is received
        QByteArray data= reply->readAll();
        emit attemptUpdateProBarValue(getTgzSize());
        file.write(data);
    });

    //use the finished signal from the reply object to close the file
    //and delete the reply object
    QObject::connect(reply, &QNetworkReply::finished, [&]{
        QByteArray data= reply->readAll();
        file.write(data);
        file.close();
        reply->deleteLater();
        event.quit();
    });

    event.exec();

    // At some point should call run again, similar to python implementation

    return 0;
}

quint64 Downloader::getTgzSize()
{
    if (datasetType == "full_set") {
        return quint64(fs::file_size(filenameAbsPath) / 100);
    } else if (datasetType == "subset") {
        return fs::file_size(filenameAbsPath);
    } else {
        return 0;
    }
}

quint64 Downloader::getTgzMax()
{
    if (datasetType == "full_set") {
        return quint64(expected_size / 100);
    } else if (datasetType == "subset") {
        return expected_size;
    } else {
        return 0;
    }
}

File: Go/CxrClassify/downloader/downloader.go
package downloader

import (
	"CxrClassify/configHandler"
	"CxrClassify/databaseHandler"
	"CxrClassify/runnable"
	"log"

	"fmt"
	"io"
	"net/http"
	"os"
)

type Downloader struct {
	runnable.Runnable

	_ func() `constructor:"init"`

	FilenameAbsPath string
	DatasetType     string
}

func (d *Downloader) init() {

}

func (d *Downloader) Setup(configHandler *configHandler.ConfigHandler, databaseHandler *databaseHandler.DatabaseHandler) {
	d.FilenameAbsPath = configHandler.GetTgzFilePath()
	d.DatasetType = configHandler.GetDatasetType()
	d.SetupRunnable(configHandler, databaseHandler)
}

func (d Downloader) Run() {
	log.Println("Checking if {} already exists")
	info, err := os.Stat(d.FilenameAbsPath)
	if err == nil && !info.IsDir() {
		log.Println("{} already exists")
		log.Println("Checking if {} was downloaded properly")
		if info.Size() == int64(d.Expected_size) {
			log.Println("{} was downloaded properly")
		} else {
			log.Println("{} was not downloaded properly")
			log.Println("Removing {}")
			e := os.Remove(d.FilenameAbsPath)
			if e != nil {
				log.Println("FAILED")
			}
			log.Println("Successfully removed {}")
			d.downloadDataset()
		}
	} else {
		log.Println("{} does not exist")
		d.downloadDataset()
	}
	d.AttemptUpdateProBarValue(1)
	// d.AttemptUpdateProBarValue(int(d.getTgzSize()))
	d.AttemptUpdateText("Image download complete")
	d.Finished()
}

func (d Downloader) downloadDataset() int {
	log.Printf("Downloading dataset from: %s", d.ConfigHandler.GetUrl())

	d.AttemptUpdateText("Downloading images")
	// d.AttemptUpdateProBarBounds(0, int(d.getTgzMax()))

	d.AttemptUpdateProBarBounds(0, 1)

	// Create the file
	out, err := os.Create(d.FilenameAbsPath)
	if err != nil {
		fmt.Println("1")
		return 1
	}
	defer out.Close()

	// Get the data
	resp, err := http.Get(d.ConfigHandler.GetUrl())
	if err != nil {
		fmt.Println("2")
		return 1
	}
	defer resp.Body.Close()

	// Check server response
	if resp.StatusCode != http.StatusOK {
		fmt.Println("3")
		return 1
	}

	// Writer the body to file
	_, err = io.Copy(out, resp.Body)
	if err != nil {
		fmt.Println("4")
		return 1
	}

	return 0
}

func (d Downloader) getTgzSize() int64 {
	fi, err := os.Stat(d.FilenameAbsPath)
	if err != nil {
		fmt.Println("5")
	}

	if d.DatasetType == "full_set" {
		return int64(fi.Size() / 100)
	} else if d.DatasetType == "subset" {
		return fi.Size()
	} else {
		return 0
	}
}

func (d Downloader) getTgzMax() int64 {
	if d.DatasetType == "full_set" {
		return int64(d.Expected_size / 100)
	} else if d.DatasetType == "subset" {
		return int64(d.Expected_size)
	} else {
		return 0
	}
}
